libcylinder/src/key/load.rs

```diff
@@ -0,0 +1,324 @@
+        None => {
+            if let Ok(user) = env::var("USER") {
+                user
+            } else {
+                whoami::username()
+            }
+        }
+    };
+
+    // if the default path is an environment variable retrieve its value
+    let key_path = match env::var(&default_path) {
+        Ok(path) => path,
+        Err(_) => default_path.to_string(),
+    };
+
+    // check if the key name is a path
+    if name.contains('/') {
+        Ok(load_key_file(None, &name)?)
+    } else {
+        // if the key name is not a path check to see if it is an environment variable
+        let name = match env::var(&name) {
+            Ok(val) => val,
+            Err(_) => name,
+        };
+        // if key_path contains multiple paths check each path
+        if key_path.contains(':') {
+            let paths = key_path.split(':');
+            for path in paths {
+                match load_key_file(Some(name.clone()), &path) {
+                    Ok(key) => return Ok(key),
+                    Err(_) => continue,
+                }
+            }
+            Err(KeyLoadError(format!(
+                "Failed to find key file in {}",
+                &key_path
+            )))
+        } else {
+            Ok(load_key_file(Some(name), &key_path)?)
+        }
+    }
+}
+
+// constructs the full path of the private key file
+fn load_key_file(key_name: Option<String>, key_path: &str) -> Result<PrivateKey, KeyLoadError> {
+    let mut path = PathBuf::from(key_path);
+    if let Some(key_name) = key_name {
+        path.push(key_name);
+    }
+    if path.exists() {
+        read_private_key(path)
+    } else {
+        path.set_extension("priv");
+        if path.exists() {
+            read_private_key(path)
+        } else {
+            Err(KeyLoadError(format!(
+                "Failed to load key: could not be found {}",
+                path.as_path().display()
+            )))
+        }
+    }
+}
+
+// Reads a private key from the given path
+fn read_private_key(path: PathBuf) -> Result<PrivateKey, KeyLoadError> {
+    let mut file = File::open(&path).map_err(|err| {
+        KeyLoadError(format!(
+            "Unable to open key file '{}': {}",
+            path.display(),
+            msg_from_io_error(err)
+        ))
+    })?;
+
+    let mut buf = String::new();
+    file.read_to_string(&mut buf).map_err(|err| {
+        KeyLoadError(format!(
+            "Unable to read key file '{}': {}",
+            path.display(),
+            msg_from_io_error(err)
+        ))
+    })?;
+    // let key = buf.trim().to_string();
+    let key = match buf.lines().next() {
+        Some(k) => k.trim().to_string(),
+        None => {
+            return Err(KeyLoadError(format!("Empty key file: {}", path.display())));
+        }
+    };
+
+    Ok(PrivateKey::new_from_hex(&key)
+        .map_err(|err| KeyLoadError(format!("unable to create private key from hex: {}", err)))?)
+}
+
+fn msg_from_io_error(err: IoError) -> String {
```

Comment:
This throws away a lot of potentially important debugging information; suggest for now we just use the display for IoError directly.

Reply:
Fixed